app/models/spree/order_contents.rb

```diff
@@ -11,25 +11,79 @@ def initialize(order)
     # Get current line item for variant if exists
     # Add variant qty to line_item
     def add(variant, quantity = 1, shipment = nil)
-      line_item = order.find_line_item_by_variant(variant)
-      add_to_line_item(line_item, variant, quantity, shipment)
+      line_item = add_to_line_item(variant, quantity, shipment)
+      update_shipment(shipment)
+      update_order
+      line_item
     end
 
     # Get current line item for variant
     # Remove variant qty from line_item
-    def remove(variant, quantity = 1, shipment = nil)
-      line_item = order.find_line_item_by_variant(variant)
+    def remove(variant, quantity = nil, shipment = nil)
+      line_item = remove_from_line_item(variant, quantity, shipment)
+      update_shipment(shipment)
+      order.update_order_fees! if order.completed?
+      update_order
+      line_item
+    end
 
-      unless line_item
-        raise ActiveRecord::RecordNotFound, "Line item not found for variant #{variant.sku}"
+    def update_or_create(variant, attributes)
```

Comment:
I wonder about the name of the method and the interface of this class. could this be update_or_add or is creating and adding different?
could we rename add to add_variant and remove to remove_variant and here we could rename to update_or_create_line ?
together with the methods taking params below this interface is quite complicated right now.

I understand the PR is basically moving it all to this one class, which is great! I am now suggesting we organize the code that has landed here :-)

Reply:
This bit wasn't from Spree. It's where I had to do some adapting to make `OrderContents` work with OFN's `CartService`: https://github.com/openfoodfoundation/openfoodnetwork/pull/7623/commits/dd34c0708f7e1fc5eeb804b81719acaec10ae4d0

it's either creating or updating a line item from a given variant, but it's a bit messy due to that extra complexity in `CartService`.

Reply:
Yeah there's a lot more that can be done here! I struggled to stop at under 30 commits, I could have carried on and done 40+.